### ISSUE TITLE
ApexSOQLInjection false positive on nested code [Apex]

### ISSUE THREAD

**Issue by ankushsomani09** [pmd/pmd#2930]
**Affects PMD Version: 6.24.0 and 6.29.0 ** 

**Rule: 	ApexSOQLInjection**

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/pmd-6.29.0/pmd_rules_apex_security.html#apexsoqlinjection>

**Description:**
Where Clause of query is build dynamically, even if i comment whole method  'buildWhereClause' which prepare where clause, still it fire the violations.

**Code Sample demonstrating the issue:**

```
/**
 *  RESTPortalGetDocuments.cls
 *
 *  @description
 *      REST Service for customer portal integration 
 *      This web service method retrieves the list of Documents for criteria specified.
 *      
 *
 *  @date - 10/OCT/2020
 *  @author - Sindhu Suru
 * 
 */
@RestResource(urlMapping='/CustomerPortalIntegrationRESTService/GetDocuments/*') 
global with sharing class RESTPortalGetDocuments {
    
    /**
     * Get Documents
     * 
     * Sample Payload:
     *  /services/apexrest/CustomerPortalIntegrationRESTService/GetDocuments?TitleOrDescription=&SolutionLine=&ProductName=&ProductVersion=&DocumentType=&DocumentStartDate=2016-10-14&DocumentEndDate=2016-10-14&Offset=0&Limit=25
     */ 
    public static String requestURI = '/services/apexrest/CustomerPortalIntegrationRESTService/GetDocuments';
    

	public static String decodeParam(String paramName) {
        String paramVal = RestContext.request.params.get(paramName);
        if(String.isBlank( paramVal )) {
            return '';
        } else {
            return EncodingUtil.urlDecode(paramVal,'UTF-8');
        }
    }
	
    @HttpGet
    global static GetDocumentsResponse GetDocuments(){  
        
        
        // RETRIEVING params
        String solutionLine = escapeQuotes( decodeParam('SolutionLine'));
        String productName = escapeQuotes( decodeParam('ProductName'));
        String prodVersion = escapeQuotes( decodeParam('ProductVersion'));
        String documentType = escapeQuotes(decodeParam('DocumentType'));
        String documentStartDate = escapeQuotes( decodeParam('DocumentStartDate'));
        String documentEndDate = escapeQuotes( decodeParam('DocumentEndDate'));
        String documentTitleOrDescription = escapeQuotes( decodeParam('TitleOrDescription'));
        String sOffSet = escapeQuotes( decodeParam('Offset'));
        String sLimit = escapeQuotes( decodeParam('Limit'));
        
        Datetime documentStartDatetime;
        Datetime documentEndDatetime;
        
        GetDocumentsResponse response;
        // VALIDATING params - user must provide either solution line or free form text to start search
        if((String.isBlank(documentTitleOrDescription) && String.isBlank(solutionLine) && String.isBlank(documentType)) ||
           String.isBlank(sOffSet) ||
           String.isBlank(sLimit)){
            return buildBadResponse(Messages.PORTAL_ERRMSG_PARAMETERS_MISSING);
        }
        
        // FORMATTING params
        if( !String.isBlank(documentStartDate) ) {
            documentStartDatetime = Date.valueOf(documentStartDate);
        }
        if( !String.isBlank(documentEndDate) ) {
            Date unformattedClose = Date.valueOf(documentEndDate);
            documentEndDatetime = DateTime.newInstanceGMT(unformattedClose, Time.newInstance(23,59,59,99));
        }
        
        Portal_Document__c pDocs = Portal_Document__c.getOrgDefaults();
        Map<Id, ContentVersion> docsVersionMap = new Map<Id, ContentVersion>();
        List<Id> matchingEntityIds =  new List<Id>();
        Set<Id> allEntityIds =  new Set<Id>(); 
        List<contentdocumentlink> linkedDocuments = new List<contentdocumentlink>();
        List<contentdocumentlink> UniqueLinkedDocuments = new List<contentdocumentlink>();
        Map<Id, Product2> matchingProducts = new Map<Id, Product2>();
        Map<Id, List<Id>> doc2ProdMap =  new Map<Id, List<Id>>();
        
        Integer lim = String.isBlank(sLimit) ? Integer.ValueOf(pDocs.Search_Record_Count__c) : Integer.ValueOf(sLimit);
        Integer offSet = String.isBlank(sOffSet) ? 0 : Integer.ValueOf(sOffSet);
        
        documentTitleOrDescription = String.isBlank( documentTitleOrDescription ) ? documentTitleOrDescription : '%'+documentTitleOrDescription+'%';

        try{
            // Getting all documents at the version level based on free form text and/or solution line
            for(ContentVersion cv : getMatchingDocuments(documentStartDatetime, documentEndDatetime, SolutionLine, DocumentType, documentTitleOrDescription, lim, offSet)){
                docsVersionMap.put(cv.contentdocumentid, cv);
            }
            
            
        }catch(Exception ex){
            return response;
        }
        return response;
     }
     
    private static List<ContentVersion> getMatchingDocuments(DateTime documentStart, DateTime documentEnd, string SolutionLine, string DocumentType, string documentTitleOrDescription, Integer lim, Integer offSet){
        
        String whereClause = buildWhereClause(documentStart, documentEnd, SolutionLine, DocumentType, documentTitleOrDescription, lim, offSet, true, true);    
        string whereEsc = escapeQuotes( whereClause);
        String query = 'select Id FROM contentversion '+whereClause ;
         System.Debug('Query = '+query);  
         return Database.query(query);                                   
     }
         
    /** 
     * To build the dynamic WHERE clause
     */
    private static String buildWhereClause(DateTime documentStart, dateTime documentEnd, string SolutionLine, string DocumentType, string documentTitleOrDescription, Integer recLIMIT, Integer recOFFSET, Boolean useOrder, Boolean isTotalRecordCount ){
        String builtWhereClause = '';
  
        if(String.IsNotBlank(documentTitleOrDescription)){
            builtWhereClause = 'WHERE (Description like :documentTitleOrDescription OR Title like :documentTitleOrDescription) ';
        }

        if(String.IsNotBlank(SolutionLine) && String.IsNotBlank(builtWhereClause)){
            builtWhereClause += 'AND Solution_Line__c INCLUDES (:SolutionLine)';
        }else if(String.IsNotBlank(SolutionLine) && String.IsBlank(builtWhereClause)){
            builtWhereClause += 'where Solution_Line__c INCLUDES (:SolutionLine)';
        }
        
        if(String.IsNotBlank(DocumentType) && String.IsNotBlank(builtWhereClause)){
            builtWhereClause += 'AND Document_Type__c =:DocumentType ';
        }else if(String.IsNotBlank(DocumentType) && String.IsBlank(builtWhereClause)){
            builtWhereClause += 'where Document_Type__c =:DocumentType ';  
        }
       
        if(documentStart != null && String.IsNotBlank(builtWhereClause)){
            builtWhereClause += 'AND (CreatedDate >=:documentStart AND CreatedDate < :documentEnd) ' ;
        }
        else if(documentStart != null && String.IsBlank(builtWhereClause)){ 
            builtWhereClause += 'where (CreatedDate >=:documentStart AND CreatedDate < :documentEnd) ' ;
        }
            
        if(String.IsNotBlank(builtWhereClause)){
            builtWhereClause += 'AND Status__c =  \'' + String.escapeSingleQuotes('Published') + '\'';
        }else if(String.IsBlank(builtWhereClause)){
            builtWhereClause += 'where Status__c = \'' + String.escapeSingleQuotes('Published') + '\'';  
        }
        
        if(String.IsNotBlank(builtWhereClause) ){
            builtWhereClause += ' AND IsLatest = true';
        }else if(String.IsBlank(builtWhereClause)){
            builtWhereClause += ' where IsLatest = true';  
        }
 
        system.debug('builtWhereClause**' + builtWhereClause);
        
        if(useOrder){
            builtWhereClause += ' ORDER BY CreatedDate DESC ';  
        } 

        return builtWhereClause;
    }
      
     private static String escapeQuotes(String s){
        if(String.isBlank(s)){
            return '';
        }else{
            return String.escapeSingleQuotes(s); 
        } 
     }

   
    
    /* Wrapper classes (response objects) */
    
    global class GetDocumentsResponse{
        public Integer TotalNumRows {set;get;} 
        public Integer RecordCnt {set;get;}
        public String ErrorMsg {set;get;}
        public List<DocumentWrapper> DocMck {set;get;}
    }
    
    global class DocumentWrapper{
        public String Id {set;}
        public String Title {set;}
        public String Description {set;}
        public String DocumentType {set;}
        public String SolutionLine {set;}
        public List<ProductWrapper> Products {set;}
    }
    
    global class ProductWrapper{
        public String Id {set;}
        public String ProductName {set;}
        public String ProductVersion {set;}
    }
}


```

**Expected outcome:**

*   Does PMD report a violation, where there shouldn't be one? -> false-positive
*   Is PMD missing to report a violation, where there should be one? -> false-negative

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by rsoesemann** [comment]
@ankushsomani09 Can you please reduce your code example to a minimum so we better and faster see what's the issue.

**Comment by ankushsomani09** [comment]
@rsoesemann : minimized the code. Please have a look. Thanks

**Comment by rsoesemann** [comment]
Still way to much. Reduce it until there are eightieth no lines anymore or no failure.

**Comment by ankushsomani09** [comment]
Reduced till last possible.

**Comment by adangel** [comment]
I've created a minimal example. I guess, solving that, is not so easy. Currently, the strings are only tracked within the method, where the database query is executed - at least, we don't have any test case, that has additional methods.

The difficulty is, to figure out, what the methods will return and don't be fooled by `RestContext.request.params.get(paramName);` - the actual parameter values are not used, only the info, whether they are present. (if they would be used directly, that would be a SOQL Injection....).

I tried out, whether I can refactor you sample. That's how it looks:

```java
public class Foo {
    public void getDocuments() {
        // RETRIEVING params
        String documentTitleOrDescription = escapeQuotes( decodeParam('TitleOrDescription'));
        String solutionLine = escapeQuotes( decodeParam('SolutionLine'));

        List<SObject> result = buildAndRunQuery(documentTitleOrDescription, solutionLine);
    }

    private List<SObject> buildAndRunQuery(String documentTitleOrDescription, String SolutionLine) {

        boolean filterByTitleDescription = String.IsNotBlank(documentTitleOrDescription);
        boolean filterBySolution = String.IsNotBlank(SolutionLine);

        String query = 'select Id FROM contentversion';

        if (filterByTitleDescription || filterBySolution) {
            query += ' WHERE';
            if (filterByTitleDescription) {
                query += ' (Description like :documentTitleOrDescription OR Title like :documentTitleOrDescription)';
            }
            if (filterByTitleDescription && filterBySolution) {
                query += ' AND';
            }
            if (filterBySolution) {
                query += ' Solution_Line__c INCLUDES (:SolutionLine)';
            }
        }
        List<SObject> res = Database.query(query);
    }

    public static String decodeParam(String paramName) {
        String paramVal = RestContext.request.params.get(paramName);
        if(String.isBlank( paramVal )) {
            return '';
        } else {
            return EncodingUtil.urlDecode(paramVal,'UTF-8');
        }
    }
}
```

Of course, the more filters you have, the messy the code becomes...
So, it would be nice, if the rule would support multiple methods.

By looking at the rule, I figured out, that it is enough to use `String.escapeSingleQuotes`, so the following line should work for you: `String whereClause = String.escapeSingleQuotes(buildWhereClause(SolutionLine, documentTitleOrDescription));`

**Comment by sindhusuru** [comment]
We are having similar issue for apex page parameters. The below code throws PMD error (ApexXSSFromURLParam) at line 2.
`String showHeaderStr = ApexPages.currentPage().getParameters().get( PARAM_SHOWHEADER );`
`if(string.isNotEmpty(showHeaderStr) && string.isNotBlank(showHeaderStr))`
`showHeaderStr = showHeaderStr.escapeHtml4();`

If we try to combine everything in one line like below then PMD passes but we will get null pointer exception if the parameter is empty. We need a way to escape and do null check at a time to pass PMD rule ApexXSSFromURLParam and also avoid null pointer exception, which is not possible. 

`String showHeaderStr = ApexPages.currentPage().getParameters().get( PARAM_SHOWHEADER ).escapeHtl4();`
			`
